enhancements/microshift/kubernetes-for-industrial-edge.md

```diff
@@ -0,0 +1,435 @@
+workloads on kubernetes. Those statements taken together lead us to
+conclude that MicroShift will only support deployments with 1 node
+acting as both control plane and worker.
+
+Because MicroShift does not reuse the system operators from OpenShift,
+it will not be configured through API resources in the way that
+OpenShift is. Instead, we will use a new configuration file for the
+control plane (`/etc/microshift/config.yaml`) with [limited
+configuration
+options](https://github.com/openshift/microshift/blob/main/pkg/config/config.go). The
+node-side configuration will be done via the standard kubelet
+configuration file.
+
+Other operating system services will be configured outside of
+MicroShift using their standard configuration files. There is no
+machine-config-operator.
+
+Because MicroShift will be treated as a single application, we will
+not expose separate health checks or implement per-component restart
+features. If any embedded component is unhealthy, the application will
+be treated as unhealthy and be restarted.
+
+### Workflow Description
+
+#### Deploying
+
+Application deployers will use RHEL for Edge tools such as Image
+Builder to construct full system images containing RHEL, MicroShift,
+agents, custom operating system configuration, data prerequisites, and
+optionally application images. Those system images will be installed
+on the target devices either individually for development/testing or
+via manufacturer imaging processes for large-scale production
+deployments. RHEL for Edge uses ostree-based images, and an [example
+workflow](https://github.com/openshift/microshift/blob/main/docs/rhel4edge_iso.md)
+for constructing a basic image is described in the MicroShift
+documentation. We are not creating a separate installer for
+MicroShift.
+
+#### Upgrading
+
+Because the operating system is based on rpm-ostree, upgrades are
+applied either by re-imaging the device, or by using the ostree update
+process and rebooting into a new ostree. The ostree images can be
+updated by downloading only the differences, which satisfies the
+requirement to minimize over-the-wire bandwidth costs of upgrades.
+
+An [example workflow illustrating
+upgrades](https://github.com/redhat-et/microshift-demos/tree/main/ostree-demo)
+is described in the MicroShift demos repository.
+
+There is no cluster-version-operator. MicroShift is not
+self-upgrading, although another agent on the host may participate in
+orchestrating upgrades.
+
+#### Configuring
+
+MicroShift reads its configuration file once on startup. Changes to
+the file will only be implemented after the application is
+restarted. Most configuration changes in production settings would be
+rolled out as new ostree images.
+
+### API Extensions
+
+MicroShift does not add any APIs to OpenShift.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+If an upgrade fails, the RHEL for Edge greenboot feature will allow
```

Comment:
can you clarify failure modes?  it doesn't extend to a system service running on rhel4edge, does it?  i assume the os can be fine, but microshift may not, and that would not necessarily result in a rollback to older?

Reply:
It depends on how the system is configured. If MicroShift has to report as healthy in order for the upgrade to be seen as successful, then yes it would rollback. That sort of decision is up to the person constructing the OS image.

Reply:
I've expanded this paragraph and linked to a blog post showing an example of green boot reverting to an older ostree based on an application running on the host.